### PR TITLE
Add support for multiple devices per SPI port.

### PR DESCRIPTION
**Pull request by willson556**
Also, utilize this capability to support multiple SPI buttons instances.

### CODE REVIEW

File: Firmware/FFBoard/Inc/SpiHandler.h
Comment (willson556) [inline Firmware/FFBoard/Inc/SpiHandler.h:26]: This is avoiding a static initialization order problem.

File: Firmware/FFBoard/Inc/cpp_target_config.h
Comment (willson556) [inline Firmware/FFBoard/Inc/cpp_target_config.h:13]: If there's a better place for these in existing files let me know!

**Comment by willson556** [comment]
I've also added support for the G27's shifter (reverse button and configurable thresholds) as discussed in Discord. Unfortunately, GitHub doesn't support stacked pull requests, so for now I've created https://github.com/willson556/OpenFFBoard/pull/1 but will open the same one here once this is reviewed.

**Comment by Ultrawipf** [comment]
Looks like a good idea for the SPI.

Give me some time to test it out if it works with the TMC and SPI DMA as the TMC heavily relies on SPI as well and should use the new system then.

A few notes:
Multiple SPI CS pins break the project on the F411 as it has only one SPI2 CS. Not too important that it works there as its mostly used for making sure code is hardware independent by having another target but the features should be disabled by the compiler if no extra CS pins are defined.

To the buttons (First glance at the code):
I am thinking about if its really a good idea to have multiple different spi button instances with different IDs (Because that could break configs and the gui) and if the shifter should really take its button from the whole button list and not manage it directly as that would break the shifter whenever you change anything in other button sources.
Its probably better to give the shifter some more modes and options to configure its pins if it does not cause big issues with the other spi and local buttons. Shifter class could have 64 different modes while we can only support 16 button sources right now.

I for example liked to disable the shifter button in the gamepad report because its not needed and can cause problems but the shifter would still need its button so its probably better to have separate modes for G29/G25 with direct SPI access or selectable pins (Which would have to depend on a target configuration again).

Maybe we should discuss a few ideas in the discord directly if you have some time because thats a very important choice for later development of the project.
I think its the right direction to manage access to the SPI ports but would prefer to implement the shifter configs a bit differently to make it easier to use.

**Comment by willson556** [comment]
> Give me some time to test it out if it works with the TMC and SPI DMA as the TMC heavily relies on SPI as well and should use the new system then.

One thing to note there is that the reconfiguration of the SPI port for every request might be a little expensive for that device. Maybe we should skip that if the device requesting the port is the same device that performed the last request.

> A few notes:
> Multiple SPI CS pins break the project on the F411 as it has only one SPI2 CS. Not too important that it works there as its mostly used for making sure code is hardware independent by having another target but the features should be disabled by the compiler if no extra CS pins are defined.

Good catch, yeah, I forgot to test out that project. That will need a little bit of thought. I think maybe statically defining the available CS pins is a bad idea and they should just be put in an array. This would imply that the CS pin for `SPIDevice`s should be user-configurable, which is _probably_ also a good idea.

> To the buttons (First glance at the code):
> I am thinking about if its really a good idea to have multiple different spi button instances with different IDs (Because that could break configs and the gui) and if the shifter should really take its button from the whole button list and not manage it directly as that would break the shifter whenever you change anything in other button sources.
> Its probably better to give the shifter some more modes and options to configure its pins if it does not cause big issues with the other spi and local buttons. Shifter class could have 64 different modes while we can only support 16 button sources right now.
> I for example liked to disable the shifter button in the gamepad report because its not needed and can cause problems but the shifter would still need its button so its probably better to have separate modes for G29/G25 with direct SPI access or selectable pins (Which would have to depend on a target configuration again).
> Then in G25/G27 Mode the shifter would handle its SPI buttons and add the additional buttons to its output as well.
> I think that makes a bit more sense than having the shifter buttons read by another spi source and passed back to the shifter just for the reverse button.

Yeah, you're probably right on having the shifter handle its SPI buttons so it appears as a single device -- that will be easier to use. Leaving the reverse buttons out of the gamepad report wasn't a use case I had considered. I can work on that.

I assume there's probably not a use-case for `SPIButtons2` outside of the G25/7 shifter at this point so I should eliminate that for now?

One question I do have is what the strategy is for GUI/Firmware compatibility. Do we expect users to keep them in sync? I personally think we should and possibly add a compatibility check. Preserving config compatibility definitely seems more important (though if that gets difficult in the future we could consider letting the GUI handle config upgrade rather than having to keep the EEPROM address map compatible).

> 
> For the G29 mode we could add just a pin selector in the gui for the local pin so its not fixed. (Would need to respect the availabel digital pins of the target)

👍 

> 
> Maybe we should discuss a few ideas in the discord directly if you have some time because thats a very important choice for later development of the project.
> I think its the right direction to manage access to the SPI ports but would prefer to implement the shifter configs a bit differently to make it easier to use.
> 
> I would merge this to a new branch or the testing branch first to integrate it into the tmc and encoder rework. For the SPI

Yeah, I'll send you a message there, definitely happy to discuss! I agree it's important to get these decisions right at this stage in the project, there's a lot more flexibility now than when there are lots of active users!

File: Firmware/FFBoard/Src/SPIButtons.cpp
Comment (Ultrawipf) [inline Firmware/FFBoard/Src/SPIButtons.cpp:19]: Would probably change the names to something more meaningful to not cause confusion with the SPI ports.
"SPI Btn 1" or "SPI Buttons 1" maybe to make it clear that the selections are just 2 independent instances instead of different SPI ports.

**Review by Ultrawipf — APPROVED** [review]
Would merge this to a testing branch if ready and then wait for the pending changes in the G27 shifter support.
Also we will confirm with testers that it all does work in a real setup and take this as a base for later development as it all seems pretty reasonable.

But we need to do some performance tests for the SPI handling if we decide to use the system for higher frequency port accesses like for encoders or motor drivers before merging to master.

File: Firmware/FFBoard/Src/SPIButtons.cpp
Comment (willson556) [reply Firmware/FFBoard/Src/SPIButtons.cpp]: 👍

**Pull request merged**
